### ISSUE TITLE
[PAID OUT] [$1000] Workspace’s avatar turns into placeholder for a second after user changes workspace in offline mode and goes online.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15627]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Login with any account.
2. Go to any workspace that doesn’t have the avatar
3. Turn off the internet
4. Try to change the avatar
5. Press Back
6. Turn on the internet connection.
7. Notice that workspace’s avatar turns into placeholder for a while after fully charged to new one.

## Expected Result:
Workspace’s avatar should change correctly after user goes from offline to online.

## Actual Result:
workspace’s avatar turns into placeholder for a while after fully charged to new one.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.78-0
**Reproducible in staging?:** y 
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/222590839-d21e08b8-25f1-490c-9580-b8fafdd3bfbe.MP4

https://user-images.githubusercontent.com/43996225/222591855-b50b2e48-b001-4381-909d-058c641f99a4.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @hungvu193 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1677776557167119

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0161188cca3cbe0452</li>
        <li>Upwork Job ID: 1633773256380624896</li>
        <li>Last Price Increase: 2023-04-05</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NicMendonca (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by zanyrenney** [comment]
Hi! reassigning this as Nicole OOO.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @zanyrenney (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by zanyrenney** [comment]
woohoo hey me!

**Comment by zanyrenney** [comment]
I can't reproduce this one on my side.

**Comment by zanyrenney** [comment]
I can't reproduce this so need to try and see if anyone else can. @NicMendonca was going to take a look too!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0161188cca3cbe0452

**Comment by MelvinBot** [comment]
Current assignee @zanyrenney is eligible for the External assigner, not assigning anyone new.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Updating the workspace avatar while offline and then going online after that will switch the avatar to the previous one for a few seconds. This is also happening to the user profile image.

### What is the root cause of that problem?

So, when we upload a new image while offline, it will queue reconnect and avatar update request. Why can there be a reconnect? The reconnect request is added because we went to the gallery to pick an image and went back to the app. Switching to online, each request will be executed sequentially, so reconnect first and then the update request.

The reconnect request returns a lot of onyx data, including the workspace/policy data which contains the old workspace avatar because our upload request hasn't been sent yet. Then, our update request will be the next to be executed and the new workspace avatar will be updated from the Pusher notification, not from the API success response. This makes the image switch between the old and the new avatar.

New -> Old -> New

The grey background appears (in step 3 above) for a glimpse because the new image is loading.

### What changes do you think we should make in order to solve the problem?


The idea of my solution is to prioritize the local avatar selected by the user. So, the prioritization will be:

1. Local avatar
2. Remote avatar
3. Default avatar

We will need a new field to save the local avatar uri. Currently, the selected avatar is saved to the `avatar` field which will be replaced by the `ReconnectApp` request, so we can't rely on that.

https://github.com/Expensify/App/blob/743bfa66f4c4c9156c2cd3897df52b6ee48863e7/src/libs/actions/Policy.js#L276-L288

I'm thinking that maybe we can use the `pendingFields` with some modifications. The new `pendingFields` will become an object with 2 properties, action & data. The action tells what action is pending while the data is the data that will be used while waiting for the data from the API. It will look like this:

```
pendingFields: {
    avatar: {
        action: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
        data: file.uri,
    },
},
```

Next, we need to also update the `successData` value to include `avatar` with the `file.uri` as the value.

https://github.com/Expensify/App/blob/743bfa66f4c4c9156c2cd3897df52b6ee48863e7/src/libs/actions/Policy.js#L289-L297

Why do we need that? Because the workspace update avatar API does not give us a response with the new avatar uploaded. We rely on the Pusher to update the new avatar. Why it matters? Let's take a look at this case below:

1. User has a workspace avatar with image A
2. User updates it to image B while offline
3. Reconnect and update avatar is queued
At this point, we have the pending avatar (`pendingFields.avatar.data`) of image B and the (not pending) avatar (`avatar`) also image B.
5. User goes online back
6. Reconnect's response update the `avatar` value to image A (the old one, but we have the pending one, so we are safe here)
7. The update request completes (but the server does not give any update response)
8. Because the request is a success, the `pendingFields` is cleared (set to null).
9. So, we left with the avatar from `avatar`.
10. A few seconds later, we receive the new avatar update from Pusher.

By setting the avatar on the `successData`, we can prevent the 9th step to happen.

The last thing we need to do is to apply the avatar prioritization at both `WorkspaceInitialPage` and `WorkspaceSettingsPage` (also on the user profile page if we want to handle it).

```
if pending avatar exist:
    return pending avatar
else if avatar exist:
   return avatar
else:
   return default avatar
```

**Comment by zanyrenney** [comment]
thanks @bernhardoj ! @MonilBhavsar let me know what you are thinking here please!

**Comment by MonilBhavsar** [comment]
@eVoloshchak could you pleas take a look at the couple of proposals. Thanks!

**Comment by eVoloshchak** [comment]
The issue is still reproducible using latest `main`

@bernhardoj's [proposal](https://github.com/Expensify/App/issues/15627#issuecomment-1461915017) is correct about the root cause: ReconnectApp is called first after app becomes online
However, I think the approach @tienifr proposing [here](https://github.com/Expensify/App/issues/15627#issuecomment-1462006535) is better: it's universal and will prevent similar issues on other screens 
We can use `_.sort` instead of finding an item and pushing it last, but that can be worked out on the PR stage

🎀👀🎀 C+ reviewed!
@MonilBhavsar

**Comment by bernhardoj** [comment]
I just want to raise my concern with moving the reconnect app to the last request. It won't work in case of these steps:
1. Turn off internet connection
2. Open Settings > Profile > Display Name
3. Change anything
4. Turn on internet connection
5. Wait until the offline indicator is hidden
6. Press Save
7. Observe the switching value still happens

The issue happens because when we reconnect, there will be only reconnect app request, so it gets executed first. Then we press save which will call the update display name api. But the reconnect request already executed (there is not a chance to reorder it), unless we cancel it and then later on, we queue it again.

Basically, it won't solve when we fire an API request while reconnect app request is ongoing. That's why I'm thinking that the data from optimistic data should be prioritized because it will be always the up-to-date data (which I think better at addressing the root cause), but I agree it's not a global solution and will require changes to all field and we should introduce a new pattern to handle this to all data.

```
if pending/optimistic data exist:
    return pending/optimistic data
else if data exist:
   return data
```

**Comment by tienifr** [comment]
> we cancel it and then later on, we queue it again.

@bernhardoj yeah in this case this is what we need to do, we cancel any pending `ReconnectApp` request by adding a new `reconnectAppCancellationController` in `HttpUtils`, and requeue the `ReconnectApp`.

**Comment by bernhardoj** [comment]
Ah, that would be great!

**Comment by MonilBhavsar** [comment]
Thanks for the proposals.
This looks like a wider problem to me and this is only instance of it. For example: the use case mentioned [here](https://github.com/Expensify/App/issues/15627#issuecomment-1466419685), and there could be more cases.

@bernhardoj I like proposal, but can we make a general or higher order fix to fix for other use-cases too, like the one you mentioned [here](https://github.com/Expensify/App/issues/15627#issuecomment-1466419685).

@tienifr I like your proposal too, but couldn't understand the follow up [here](https://github.com/Expensify/App/issues/15627#issuecomment-1466506117). Could you please clarify how are you looking to cancel Reconnect. In that case, Reconnect has already been finished, unless we trigger simultaneously.

**Comment by tienifr** [comment]
> In that case, Reconnect has already been finished, unless we trigger simultaneously.

@MonilBhavsar In that case, the `ReconnectApp` request is still ongoing, if it has been finished before the name update request, it will not cause any issue since it will not override any optimistic data.

> Could you please clarify how are you looking to cancel Reconnect

Here's what will happen:
1. The user presses 'Save'
2. The name save request is added to the `persistedRequests` list, at this moment we'll cancel all the `pending` `ReconnectApp` request
This can be done by (this is similar to the generic `cancellationController` which we already have for all requests):
- Add a `let reconnectAppCancellationController = new AbortController();` in `src/libs/HttpUtils.js`
- In `processHTTPRequest`, If the request is a `ReconnectApp` request, set the signal of the request to `reconnectAppCancellationController.signal`
- Add a new method
```
function cancelPendingReconnectAppRequests() {
    reconnectAppCancellationController.abort();
    reconnectAppCancellationController = new AbortController();
}
```
- Call the `cancelPendingReconnectAppRequests` to cancel the `ReconnectApp` request
3. After the `ReconnectApp` is cancelled, we'll add the name save request, then we'll queue the `ReconnectApp` again to the `persistedRequests`. Now `ReconnectApp` is the last request in the queue, just as we expect it.

This will ensure that it will work correctly even on the very edge case mentioned [here](https://github.com/Expensify/App/issues/15627#issuecomment-1466419685).

**Comment by bernhardoj** [comment]
@MonilBhavsar Thanks for taking a look. My proposal can't become a general solution to fix all cases because it requires to change (the changes is at my proposal) all optimistic data we have.

**Comment by MonilBhavsar** [comment]
Discussed internally and we think we can fix a wider problem with @tienifr 's propsal. @eVoloshchak what do you think?

**Comment by eVoloshchak** [comment]
> Discussed internally and we think we can fix a wider problem with @tienifr 's propsal. @eVoloshchak what do you think?

I agree, I think we should go with @tienifr's proposal. It will prevent similar issues app-wide

**Comment by MelvinBot** [comment]
@eVoloshchak @MonilBhavsar @zanyrenney this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MonilBhavsar** [comment]
@tienifr let's fix this with your proposal.
Please make sure to test well as we're updating very higher order and basic file

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @MonilBhavsar!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
The PR is ready for review [#16093 ](https://github.com/Expensify/App/pull/16093). Thanks

**Comment by tienifr** [comment]
@MonilBhavsar @eVoloshchak How's it going? Could you please take a look at the [PR](https://github.com/Expensify/App/pull/16093)?

**Comment by eVoloshchak** [comment]
Reviewing shortly

**Comment by tienifr** [comment]
# Regression Test Proposal

**Bug**: Workspace’s avatar turns into placeholder for a second after user changes workspace in offline mode and goes online

**Proposed Test Steps**:

1. Login with any account.
2. Go to any workspace that doesn’t have the avatar
3. Turn off the internet
4. Change the avatar, after that press Back
5. Turn on the internet connection
6. Verify that workspace’s avatar does not turn into placeholder for a while after fully changes to the new one

[video](https://user-images.githubusercontent.com/113963320/226160955-6cf95806-a884-465d-a913-510386ab49db.mp4)

Do we :+1: or :-1:

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.89-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16093


If no regressions arise, payment will be issued on 2023-04-03. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@eVoloshchak / @MonilBhavsar] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@eVoloshchak / @MonilBhavsar] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@eVoloshchak / @MonilBhavsar] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@zanyrenney] Determine if we should create a regression test for this bug.
- [ ] [@eVoloshchak] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@zanyrenney] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by zanyrenney** [comment]
hey @MonilBhavsar @eVoloshchak please can you complete the tasks from the checklist above so I can complete mine and pay this out. Thanks!

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: I don't think this is applicable here, since that's how it was implemented initially. And since this was happening on many other screens (essentially every change you did while offline), there is no PR I could pinpoint 

- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A

- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: https://expensify.slack.com/archives/C049HHMV9SM/p1680712280218859

**Comment by zanyrenney** [comment]
https://expensify.slack.com/archives/C049HHMV9SM/p1678982415645269

Shouldn't hacve been doubled. Changing price to $1000.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by eVoloshchak** [comment]
Not overdue, this is awaiting for BugZero Checklist to be finished and payments sorted

**Comment by zanyrenney** [comment]
@hungvu193 invited to job as reporter
@tienifr invited as contributor 
@eVoloshchak invited as contributor+ 

Please can you all accept the offers so we can pay this out. In future. it's also much easier if you apply to the job directly! thanks!

**Comment by eVoloshchak** [comment]
Not overdue, this is awaiting for BugZero Checklist to be finished

**Comment by zanyrenney** [comment]
All have been paid out!

**Comment by MonilBhavsar** [comment]
@eVoloshchak could you please propose a regression test here

**Comment by eVoloshchak** [comment]
**Regression Test Proposal**

1. Go to any workspace that doesn’t have an avatar
2. Turn off the internet
3. Click on the Workspace avatar and change it to any picture
4. Navigate back
5. Turn on the internet
6. Verify the workspace avatar is not changed to a placeholder image

**Do we agree 👍 or 👎**

**Comment by MonilBhavsar** [comment]
> **Regression Test Proposal**
> 
> 1. Go to any workspace that doesn’t have an avatar
> 2. Turn off the internet
> 3. Click on the Workspace avatar and change it to any picture
> 4. Navigate back
> 5. Turn on the internet
> 6. Verify the workspace avatar is not changed to a placeholder image
> 
> **Do we agree 👍 or 👎**

👍 Looks good, some minor changes. cc @zanyrenney 

2. Go offline
3. Click on the Workspace avatar and upload any image
5. Go online
6. Verify the workspace avatar is not changed to a default/placeholder image

**Comment by zanyrenney** [comment]
Oh nice, those changes are great. Thanks @MonilBhavsar

**Comment by MonilBhavsar** [comment]
I updated the steps in the new issue as github messed with the numbering 😂 


https://user-images.githubusercontent.com/32012005/233084196-609f7b35-72af-4339-990a-e8787fc3f5ba.mov

**Comment by zanyrenney** [comment]
Completed, and regression test requested.

**Comment by marcaaron** [comment]
Sorry everyone, we are going to need to find a new solution to this issue as I am reverting the code that we shipped to fix this issue. I commented last month [here](https://github.com/Expensify/App/pull/16093#pullrequestreview-1424890391) that the solution was not really appropriate, but no action was taken on that and it ended up blocking a critical performance improvement so now I'm reverting it.

The PR where I am reverting it is [here](https://github.com/Expensify/App/pull/18807).

Sorry about this and thanks for your patience and understanding.

**Comment by zanyrenney** [comment]
Hey @marcaaron can you clarify what the steps needed here then are? Should we be recreating another Upwork job and looking for alternative solutions to that? Thanks!

**Comment by marcaaron** [comment]
I am not sure _what_ we want to do here. But the original solution is ❌

What makes the most sense to me is to keep this issue open so we can take another crack at solving it at it's "root cause".

I'd probably try to start some discussion internally before creating another Upwork job to make sure it's something we want to field proposals for. I can't say for sure - but some more care and thought needs to go into fixing this bug. 

Maybe @MonilBhavsar can lead this up? IIRC we previously discussed this PR in Slack.

**Comment by zanyrenney** [comment]
Thanks @marcaaron that's super helpful. I appreciate the additional context.

**Comment by zanyrenney** [comment]
@MonilBhavsar do you have capacity to lead up the discussion about if we want to fix this?

**Comment by MonilBhavsar** [comment]
Yes, can we please confirm the issue is still reproducible. Would be good to confirm.

**Comment by zanyrenney** [comment]
Hey, I am heading OOO shortly, so assigning another bug0 team member to help watch over this issue whilst I am OOO. The latest is from @MonilBhavsar above, where we are wanting to check reproducibility.

**Comment by zanyrenney** [comment]
hey gal friend, please can you watch over this one? No payouts required ( I have done them all already) just to check the reproducibility. @maddylewis

**Comment by maddylewis** [comment]
i cannot reproduce this and/or the result i got looks like expected behavior. the icon is slightly greyed out for a short time while the internet reconnects - 

https://github.com/Expensify/App/assets/38016013/c492c9ac-a276-4b14-adaa-8cba89095bc1

**Comment by marcaaron** [comment]
Sorry - this is my bad. It will be reproducible once we rollback the code. But is still in reviews so that's why it is not reproducible.

Let's maybe close this and I will make a note to either reopen it or create a new issue and link back to this one so we can restart the process. We didn't quite get it right this time - but that's OK - we will try again! 💪
